rfcs/20190910-struct-tensor.md

```diff
@@ -0,0 +1,885 @@
+        {'amount': 1.0, 'unit': 'cup', 'name': 'butter'},
+        {'amount': 1.0, 'unit': 'teaspoon', 'name': 'cinnamon'},
+        {'amount': 2.0, 'unit': 'teaspoon', 'name': 'cream of tartar'},
+         ...],
+    'step': [
+         {'description': 'Preheat oven to 350 degrees F.'}
+         {'description': 'Whisk together the flour, cinnamon, baking soda, salt, ...'}
+         {'description': 'Using an electric mixer, cream together the butter and sugar ...'}
+         {'description': 'Add the dry ingredients to the wet ingredients and mix until ...'}
+         ...],
+    'user_rating': [
+         {'user_embedding': [0.7, 2.0, 0.3, 0.3, 5.2, 2.2], score: 0.8},
+         {'user_embedding': [1.4, 0.0, 3.1, 1.1, 1.2, 0.3], score: 0.4},
+         ...],
+   })
+```
+
+In this example:
+
+* The root object `struct` is a scalar `StructTensor` (i.e., a single
+  structure).
+* The `recipe` field contains a scalar `StructTensor` (a single nested
+  structure).
+* The nested fields `ingredients`, `step`, and `user_rating` contain vector
+  `StructTensor`s (i.e. 1D collections of structures with the same schema).
+* The nested fields `title`, `est_time`, `amount`, `unit`, `name`,
+  `description`, and `score` contain scalar `Tensor`s.
+* The nested `user_embedding` field contains a vector `Tensor`.
+
+In the initial implementation, the value for a field may be a `Tensor`, a
+`RaggedTensor`, or a nested `StructTensor`, where nested `StructTensor`s could
+contain single structures (scalar `StructTensor`) or collections of structures
+(non-scalar `StructTensor`).  In the future, we may add support for additional
+value types, such as `SparseTensor`s.
+
+A `StructTensor`'s schema constrains the type for each field:
+
+* **For fields with `Tensor` values**, the schema specifies the field value's
+  `dtype` and `shape`.
+* **For fields with `RaggedTensor` values**, the schema specifies the field
+  value's `dtype` and `ragged_rank`.
+* **For fields with `StructTensor` values**, the schema specifies the nested
+  `StructTensor` value's `shape` (e.g., a single structure vs a list of
+  structures), and the name and type for each nested field.
+
+For multidimensional `StructTensor` (such as a vector of structures), the
+individual structures must all share the same schema.
+
+### Usage
+
+#### Construction
+
+`StructTensor`s can be constructed from a variety of sources, including:
+
+* Nested Python dictionaries and lists.
+* Serialized protocol buffer messages.
+* JSON-encoded protocol buffers.
+* Apache Arrow record batches.
+* Apache Parquet.
+* Pandas DataFrames.
+* Spanner.
+* Firebase.
+
+For more details, see the section below on "Integration with Other Formats."
+`StructTensor`s may also be constructed directly from a shape and a field
+dictionary; see the "Struct Tensor Encoding" section below for more details.
+
+#### Tensorflow Ops
+
+We will extend standard Tensorflow array operations to support `StructTensor`s,
```

Comment:
At the design review, we decided not to add operator dispatch for StructTensor for array ops at this time.  They may be added in the future, if there is demand for them.